File: 2026-04/2026-04-23/closest_time_direction.py
import datetime as dt
SECONDS_IN_HALF_DAY = 43200

def get_direction(time1, time2):
    dummy_date = dt.date(1, 1, 1)
    time1_datetime_object = dt.datetime.combine(dummy_date, dt.time.fromisoformat(time1))
    time2_datetime_object = dt.datetime.combine(dummy_date, dt.time.fromisoformat(time2))

    difference = time2_datetime_object - time1_datetime_object

    return "forward" if difference.seconds < SECONDS_IN_HALF_DAY \
        else "backward" if difference.seconds > SECONDS_IN_HALF_DAY \
        else "equal"

# --- TEST SUITE ---

tests_text = r'''
1. get_direction("10:00", "12:00") should return "forward".
2. get_direction("11:00", "05:00") should return "backward".
3. get_direction("00:00", "12:00") should return "equal".
4. get_direction("15:45", "01:10") should return "forward".
5. get_direction("03:30", "19:50") should return "backward".
6. get_direction("06:30", "18:30") should return "equal".
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("——————————————————————————",
        "\n🧪 Starting Verification...",
        "\n——————————————————————————")

    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            fail_count += 1

        print("————————————————————————————")

    print(
        f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED." if fail_count
        else "🎉 SUCCESS: All tests PASSED."
    )

run_tests(test_data)

File: 2026-04/2026-04-23/closestTimeDirection.js
function getDirection(time1, time2) {
  const time1Temporal = Temporal.PlainTime.from(time1);
  const time2Temporal = Temporal.PlainTime.from(time2);

  const difference = time1Temporal.until(time2Temporal);

  if (Math.abs(difference.hours) === 12 && difference.minutes === 0) {
    return "equal";
  }

  if (0 <= difference.hours && difference.hours < 12 || -23 <= difference.hours && difference.hours <= -12) {
    return "forward";
  }

  if (12 <= difference.hours && difference.hours <= 23 || -12 < difference.hours && difference.hours <= 0) {
    return "backward";
  }
}

// --- TEST SUITE ---

const testsText = String.raw`
1. getDirection("10:00", "12:00") should return "forward".
2. getDirection("11:00", "05:00") should return "backward".
3. getDirection("00:00", "12:00") should return "equal".
4. getDirection("15:45", "01:10") should return "forward".
5. getDirection("03:30", "19:50") should return "backward".
6. getDirection("06:30", "18:30") should return "equal".
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("——————————————————————————",
            "\n🧪Starting Verification...",
            "\n——————————————————————————");

  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);

    const comparison = Array.isArray(testOutput)
      ? arraysEqual(functionCallOutput, testOutput)
      : functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    } else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      ++failCount;
    }
    console.log("————————————————————————————");
  })

  console.log(failCount
    ? `⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`
    : "🎉SUCCESS: All tests PASSED."
  );
}

function arraysEqual(a, b) {
  if (a.length !== b.length) return false;

  for (let i = 0; i < a.length; ++i) {
    if (Array.isArray(a[i])) {
      if (Array.isArray(b[i])) {
        if (!arraysEqual(a[i], b[i])) {
          return false;
        }
      } else {
        return false;
      }
    } else if (a[i] !== b[i]) {
      return false;
    }
  }

  return true;
}

runTests(testData);